base/mapiterator.jl

```diff
@@ -0,0 +1,249 @@
+# available.
+#
+#   collect(StreamMapItr(f, c...)) == map(f, c...)
+#
+# The maximum number of concurrent @async tasks can be set with the
+# "ntasks" option. e.g.
+#
+#    StreamMapItr(f, c...; ntasks=10)
+#
+#-------------------------------------------------------------------------------
+
+
+type StreamMapItr
+    async_itr::AsyncMapItr
+end
+
+function StreamMapItr(f, c...; ntasks=nothing)
+    StreamMapItr(AsyncMapItr(f, Dict{Int,Any}(), c...; ntasks=ntasks))
+end
+
+
+type StreamMapState
+    i::Int
+    async_state::AsyncMapState
+end
+
+
+start(itr::StreamMapItr) = StreamMapState(0, start(itr.async_itr))
+
+# Done when source async iterator is done and all results have been consumed.
+function done(itr::StreamMapItr, state::StreamMapState)
+    done(itr.async_itr, state.async_state) && isempty(itr.async_itr.results)
+end
+
+
+# Pump the source async iterator if it is not already busy...
+
+function pump_source(itr::StreamMapItr, state::StreamMapState)
+    if !isbusy(itr.async_itr, state.async_state) &&
+       !done(itr.async_itr, state.async_state)
+        ignored, state.async_state = next(itr.async_itr, state.async_state)
+        return true
+    else
+        return false
+    end
+end
+
+function next(itr::StreamMapItr, state::StreamMapState)
+
+    state.i += 1
+
+    results = itr.async_itr.results
+    while !haskey(results, state.i)
+
+        # Wait for results to become available...
+        if !pump_source(itr,state) && !haskey(results, state.i)
+            wait(state.async_state)
+        end
+    end
+    r = results[state.i]
+    delete!(results, state.i)
+
+    return (r, state)
+end
+
+
+
+#-------------------------------------------------------------------------------
+# Interface: amap and imap
+#-------------------------------------------------------------------------------
+
+"""
+    amap(f, c...; ntasks=100) -> collection
+
+Transform collection c by applying f to each element using at most
+100 asynchronous tasks. For multiple collection arguments, apply f
+elementwise.
+Note: `amap(f, c...; ntasks=1)` is equivalent to `map(f, c...)`.
+"""
+amap(f, c...; kv...) = collect(imap(f, c...; kv...))
+
+
+"""
+    imap(f, c...; ntasks=100) -> iterator
+
+Apply f to each element of c using at most 100 asynchronous tasks.
+For multiple collection arguments, apply f elementwise.
+Note: `collect(imap(f, c...; ntasks=1))` is equivalent to `map(f, c...)`.
+"""
+imap(f, c...; ntasks=nothing) = StreamMapItr(f, c...; ntasks=ntasks)
+
+
+"""
+    amap!(function, collection; ntasks=100)
+
+In-place version of [`amap`](:func:`amap`).
+"""
+function amap! end
+
+"""
+    amap!(function, destination, collection...; ntasks=100)
+
+Like [`amap`](:func:`amap`), but stores the result in `destination` rather than a new collection.
+`destination` must be at least as large as the first collection.
+"""
+function amap!(f, c...; ntasks=nothing)
+
+    destination = c[1]
```

Comment:
The signature shouldn't use splatting in this way, providing only one input should be a `MethodError` not a `BoundsError`